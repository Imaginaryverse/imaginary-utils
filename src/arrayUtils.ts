export function sortByProperty<T extends object>(
  arr: T[],
  accessor: (item: T) => string | number | Date,
  direction: 'asc' | 'desc'
): T[] {
  if (!Array.isArray(arr)) {
    throw new Error('arr must be an array');
  }
  if (typeof accessor !== 'function') {
    throw new Error('accessor must be a function');
  }

  if (
    typeof accessor(arr[0]) !== 'string' &&
    typeof accessor(arr[0]) !== 'number' &&
    !(accessor(arr[0]) instanceof Date)
  ) {
    throw new Error('accessor must return a string, number, or Date');
  }

  return arr.sort((a, b) => {
    const aVal = accessor(a);
    const bVal = accessor(b);
    if (direction === 'asc') {
      if (aVal < bVal) {
        return -1;
      }
      if (aVal > bVal) {
        return 1;
      }
      return 0;
    } else {
      if (aVal > bVal) {
        return -1;
      }
      if (aVal < bVal) {
        return 1;
      }
      return 0;
    }
  });
}
