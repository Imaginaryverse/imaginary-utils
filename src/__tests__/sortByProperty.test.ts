import { sortByProperty } from '../arrayUtils';

const array = [
  { name: 'John', id: 1, birthdate: new Date(1990, 0, 1) },
  { name: 'Jane', id: 2, birthdate: new Date(1995, 0, 1) },
  { name: 'Joe', id: 3, birthdate: new Date(1993, 0, 1) },
];

describe('sortByProperty', () => {
  describe("when given an array of objects, an accessor function, and a direction of 'asc'", () => {
    it('returns a sorted array', () => {
      expect(sortByProperty(array, item => item.name, 'asc')).toMatchSnapshot();
      expect(sortByProperty(array, item => item.id, 'asc')).toMatchSnapshot();
      expect(
        sortByProperty(array, item => item.birthdate, 'asc')
      ).toMatchSnapshot();
    });
  });

  describe("when given an array of objects, an accessor function, and a direction of 'desc'", () => {
    it('returns a sorted array', () => {
      expect(
        sortByProperty(array, item => item.name, 'desc')
      ).toMatchSnapshot();
      expect(sortByProperty(array, item => item.id, 'desc')).toMatchSnapshot();
      expect(
        sortByProperty(array, item => item.birthdate, 'desc')
      ).toMatchSnapshot();
    });
  });

  describe('when given an invalid array', () => {
    it('throws an error', () => {
      expect(() =>
        sortByProperty(null as any, item => item as any, 'asc')
      ).toThrow('arr must be an array');
    });
  });

  describe('when given an invalid accessor', () => {
    it('throws an error', () => {
      expect(() => sortByProperty(array, null as any, 'asc')).toThrow(
        'accessor must be a function'
      );
    });
  });

  describe('when given an accessor that returns an invalid type', () => {
    it('throws an error', () => {
      expect(() => sortByProperty(array, item => item as any, 'asc')).toThrow(
        'accessor must return a string, number, or Date'
      );
    });
  });
});
