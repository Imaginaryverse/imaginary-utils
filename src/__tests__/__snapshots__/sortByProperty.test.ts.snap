// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortByProperty when given an array of objects, an accessor function, and a direction of 'asc' returns a sorted array 1`] = `
[
  {
    "birthdate": 1994-12-31T23:00:00.000Z,
    "id": 2,
    "name": "Jane",
  },
  {
    "birthdate": 1992-12-31T23:00:00.000Z,
    "id": 3,
    "name": "Joe",
  },
  {
    "birthdate": 1989-12-31T23:00:00.000Z,
    "id": 1,
    "name": "John",
  },
]
`;

exports[`sortByProperty when given an array of objects, an accessor function, and a direction of 'asc' returns a sorted array 2`] = `
[
  {
    "birthdate": 1989-12-31T23:00:00.000Z,
    "id": 1,
    "name": "John",
  },
  {
    "birthdate": 1994-12-31T23:00:00.000Z,
    "id": 2,
    "name": "Jane",
  },
  {
    "birthdate": 1992-12-31T23:00:00.000Z,
    "id": 3,
    "name": "Joe",
  },
]
`;

exports[`sortByProperty when given an array of objects, an accessor function, and a direction of 'asc' returns a sorted array 3`] = `
[
  {
    "birthdate": 1989-12-31T23:00:00.000Z,
    "id": 1,
    "name": "John",
  },
  {
    "birthdate": 1992-12-31T23:00:00.000Z,
    "id": 3,
    "name": "Joe",
  },
  {
    "birthdate": 1994-12-31T23:00:00.000Z,
    "id": 2,
    "name": "Jane",
  },
]
`;

exports[`sortByProperty when given an array of objects, an accessor function, and a direction of 'desc' returns a sorted array 1`] = `
[
  {
    "birthdate": 1989-12-31T23:00:00.000Z,
    "id": 1,
    "name": "John",
  },
  {
    "birthdate": 1992-12-31T23:00:00.000Z,
    "id": 3,
    "name": "Joe",
  },
  {
    "birthdate": 1994-12-31T23:00:00.000Z,
    "id": 2,
    "name": "Jane",
  },
]
`;

exports[`sortByProperty when given an array of objects, an accessor function, and a direction of 'desc' returns a sorted array 2`] = `
[
  {
    "birthdate": 1992-12-31T23:00:00.000Z,
    "id": 3,
    "name": "Joe",
  },
  {
    "birthdate": 1994-12-31T23:00:00.000Z,
    "id": 2,
    "name": "Jane",
  },
  {
    "birthdate": 1989-12-31T23:00:00.000Z,
    "id": 1,
    "name": "John",
  },
]
`;

exports[`sortByProperty when given an array of objects, an accessor function, and a direction of 'desc' returns a sorted array 3`] = `
[
  {
    "birthdate": 1994-12-31T23:00:00.000Z,
    "id": 2,
    "name": "Jane",
  },
  {
    "birthdate": 1992-12-31T23:00:00.000Z,
    "id": 3,
    "name": "Joe",
  },
  {
    "birthdate": 1989-12-31T23:00:00.000Z,
    "id": 1,
    "name": "John",
  },
]
`;
